### PR TITLE
chore: run-history v3

### PR DESCRIPTION
**Pull request by tssweeney**
This PR introduces a new run-history op: `run-history3` which is an optimized run history focused on parquet and stream tables. It is general purpose (other than legacy media) and leverages much of the underpinning of `history2`. Currently, it is only exposed in `StreamTable.rows`, but I expect after sussing out any bugs, we can replace all run-history with this op.

History2 was a huge improvement over the first implementation as it moved us to livesets & parquet and laid the groundwork for column-level translation. History3 builds on this model, and adds additional support for:
* Any arbitrary shape is now supported. This means any combination of nested list/dicts/unions/custom types/custom objects now works.
    * Note: you must log weave objects using the Weave Stream Table / Lite Run logging for the weave types/objects to work
    * Legacy Media is not currently supported
* Generalized column pushdown. Now we can use any combination of dot separated pick operations and we will get the correct data back out - properly unnesting history stores nested objects
* Minimal non-vectorized transformations. Now we can transform gorilla parquet history into ArrowWeaveList using vector options except for: a) complex unions; b) Legacy media types.
     * Note: there are paths forward here, but I need to land this and move to the next step for now.

Note: These tests rely on using a custom local-container until https://github.com/wandb/core/pull/15838 lands.

### CODE REVIEW

File: weave/ops_domain/run_history/run_history_v3_parquet_weave_only.py [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+
+@dataclasses.dataclass
+class PathTree:
+    children: typing.Dict[PathItemType, "PathTree"] = dataclasses.field(
+        default_factory=dict
+    )
+    data: typing.Optional[typing.Any] = None
+
+
+def _get_history_stream(run: wdt.Run, columns=None):
```
Comment (tssweeney) [inline weave/ops_domain/run_history/run_history_v3_parquet_weave_only.py:91]: Pretty much this entire PR is about this function

**Review by shawnlewis — APPROVED** [review]

File: weave/ops_arrow/convert.py [hunk 2]
```diff
@@ -289,7 +294,10 @@ def to_arrow_from_list_and_artifact(
 
 
 def to_arrow(
-    obj, wb_type=None, artifact: typing.Optional[artifact_base.Artifact] = None
+    obj,
+    wb_type=None,
+    artifact: typing.Optional[artifact_base.Artifact] = None,
+    py_objs_already_mapped: bool = False,
```
Comment (dannygoldstein) [inline weave/ops_arrow/convert.py:300]: hmm... this feels like a smell. why would we need to call `to_arrow` if the objects are already mapped?

File: weave/ops_domain/run_history/history_op_common.py [hunk 3]
```diff
@@ -110,22 +117,103 @@ def get_top_level_keys(key_tree: KeyTree) -> list[str]:
 def get_full_columns(columns: typing.Optional[list[str]]):
     if columns is None:
         return None
-    return list(set([*columns, "_step", "_runtime", "_timestamp"]))
+    return list(set([*columns, "_step"]))
```
Comment (dannygoldstein) [inline weave/ops_domain/run_history/history_op_common.py:120]: aren't `_runtime` and `_timestamp` still going to be present for history1&2?

File: weave/ops_domain/run_history/history_op_common.py [hunk 4]
```diff
@@ -197,48 +291,73 @@ def history_keys(run: wdt.Run) -> list[str]:
...
-                span.set_tag("path", path)
-                table = pq.read_table(path, columns=columns_to_read)
-
-            # convert table to ArrowWeaveList
-            with tracer.trace("make_awl") as span:
-                awl: ArrowWeaveList = ArrowWeaveList(table, object_type=object_type)
-            tables.append(awl)
-    list = make_list(**{str(i): table for i, table in enumerate(tables)})
-    concatted = use(concat(list))
+def local_path_to_parquet_table(
```
Comment (dannygoldstein) [inline weave/ops_domain/run_history/history_op_common.py:294]: nit: this is returning an AWL not a parquet table / arrow table, maybe something like 
```suggestion
def awl_from_local_parquet_path(
```

File: weave/ops_arrow/convert.py [hunk 5]
```diff
@@ -150,6 +150,8 @@ def none_unboxer(iterator: typing.Iterable):
                     if py_obj is None:
                         data.append(None)
                     elif py_objs_already_mapped:
+                        if isinstance(py_obj, dict) and "_val" in py_obj:
+                            py_obj = py_obj["_val"]
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/ops_arrow/convert.py:154]: When we encode Weave types in history, we basically do a storage.to_python on it, which results in a `_type` and `_val` field. The rest of the system expects only the `_val` component. This is pretty much the main special case.

File: weave/ops_arrow/convert.py [hunk 6]
```diff
@@ -250,7 +252,10 @@ def none_unboxer(iterator: typing.Iterable):
         )
 
     if py_objs_already_mapped:
-        return pa.array(py_objs, pyarrow_type)
+        return pa.array(
+            [p["_val"] if isinstance(p, dict) and "_val" in p else p for p in py_objs],
+            pyarrow_type,
+        )
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/ops_arrow/convert.py:258]: see comment above

File: weave/ops_domain/run_history/history_op_common.py (hunk 3 above)
Comment (tssweeney) [reply weave/ops_domain/run_history/history_op_common.py:120]: yes, but `_step` is the only key that internal resolver logic depends on. The others are fine to be selected by the user if desired

File: weave/ops_domain/run_history/history_op_common.py (hunk 4 above)
Comment (tssweeney) [reply weave/ops_domain/run_history/history_op_common.py:294]: nice catch, fixing

File: weave/ops_arrow/convert.py (hunk 2 above)
Comment (tssweeney) [reply weave/ops_arrow/convert.py:300]: hmmm ... i see your point. This relates to the above comments. Upon reflection, maybe it is better to do the _val stuff externally, then directly construct the awl. I'll revisit this today

Comment (tssweeney) [reply weave/ops_arrow/convert.py:300]: hmmm unions are sort of nasty here... we still want the mappers to handle this

**Review by dannygoldstein — APPROVED** [review]

**Pull request merged**